ufl/algorithms/map_integrands.py

```diff
@@ -35,7 +37,23 @@ def map_integrands(function, form, only_integral_type=None):
             return itg.reconstruct(function(itg.integrand()))
         else:
             return itg
-    elif isinstance(form, Expr):
+    elif isinstance(form, FormSum):
+        mapped_components = [map_integrands(function, component, only_integral_type)
+                             for component in form.components()]
+        nonzero_components = [(component, 1) for component in mapped_components
+                              # Catch ufl.Zero and zeros of dual objects
+                              # Workaround while the role of `__eq__`/`equals` for BaseForms is clarified ?
+                              if not component == 0]
+        return FormSum(*nonzero_components)
+    elif isinstance(form, Adjoint):
+        # Zeros are caught inside `Adjoint.__new__`
+        return Adjoint(map_integrands(function, form._form, only_integral_type))
+    elif isinstance(form, Action):
+        left = map_integrands(function, form._left, only_integral_type)
+        right = map_integrands(function, form._right, only_integral_type)
+        # Zeros are caught inside `Action.__new__`
+        return Action(left, right)
+    elif isinstance(form, (Expr, BaseForm)):
```

Comment:
@wence- @dham : Following our discussion regarding the fact that the derivative expansion was manually pushed through in `expand_derivatives`. I have made some cleanup and added handlers here to address the issue raised but also to extend this behaviour to other MultiFunction objects that could eventually be fed with dual objects (cf. this commit https://github.com/firedrakeproject/ufl/commit/2fdd18095ed48342080ad73ddc7fb41c7ad9fa30)